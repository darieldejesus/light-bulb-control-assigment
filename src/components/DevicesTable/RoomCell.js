import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import {
  Box,
  Txt,
  Button,
  Flex,
  Input,
} from 'rendition';
import FaPencil from 'react-icons/lib/fa/pencil';
import FaDiscard from 'react-icons/lib/fa/mail-reply';
import FaCheck from 'react-icons/lib/fa/floppy-o';
import FaUpdating from 'react-icons/lib/fa/spinner';
import styled from 'styled-components';

import { updateName } from '../../stateManager/ducks/devices';

const CellContainer = styled(Box)`
  position: relative;
`;

const ActionButton = styled(Button)`
  display: ${({ show }) => show ? 'inline-block' : 'none'}
`;

const EditButton = ActionButton.extend`
  position: absolute;
  left: -1.8em;
  top: 0.1em;
  
  &:hover, :active {
    color: unset;
  }
`;

const mapStateToProps = ({ devices }) => ({
  currentDevice: devices.currentDevice,
  isUpdating: devices.isUpdatingName,

});

const mapDispatchToProps = dispatch => ({
  updateDeviceName: (id, device) => dispatch(updateName(id, device)),
});

export class RoomCell extends Component {
  constructor(props) {
    super(props);
    this.state = {
      edit: false,
    };
  }

  /**
   * Prevent component to be re-rendered if isUpdating is active
   * but the component is not being edited.
   *
   * @param {Object} nextProps - Next component props.
   * @param {boolean} nextProps.isUpdating - Tells if a component is being updated.
   * @param {Object} nextState - Next component state.
   * @param {boolean} nextState.show - Tells if a component is selected.
   * @returns {boolean} False or True depending on the condition.
   */
  shouldComponentUpdate({ isUpdating }, { show }) {
    const { isUpdating: currentUpdating } = this.props;
    return !(isUpdating !== currentUpdating && !show);
  }

  static getDerivedStateFromProps({ value, currentDevice }, { name, edit }) {
    const { id, name: deviceName } = currentDevice;
    const show = value === deviceName;
    return {
      id,
      show,
      name: edit && show ? name : value,
      edit: edit && !show ? false : edit,
    };
  }

  handleOnToggleView = () => {
    this.setState(({ edit }) => ({
      edit: !edit,
    }));
  };

  handleOnChangeName = (e) => {
    this.setState({
      name: e.target.value,
    });
  };

  handleOnEditClick = () => {
    const { updateDeviceName } = this.props;
    const { id, name } = this.state;
    updateDeviceName(id, {
      name,
    });
  };

  render() {
    const { show, edit, name } = this.state;
    const { isUpdating } = this.props;
    const tooltip = edit && show
      ? 'Cancel edition'
      : 'Edit Room name';
    const icon = edit && show
      ? <FaDiscard />
      : <FaPencil />;

    return (
      <CellContainer>
        <Flex>
          <EditButton
            tooltip={tooltip}
            danger={edit}
            m={2}
            fontSize={[3, 2]}
            show={show}
            plaintext
            primary
            onClick={this.handleOnToggleView}
            iconElement={icon}
          />
          {edit && show ? ([
            <Input
              key="EditInput"
              value={name}
              autoFocus
              fontSize={0}
              width="8em"
              px={1}
              placeholder="Room name"
              onChange={this.handleOnChangeName}
            />,
            !isUpdating && (
              <ActionButton
                key="SaveChanges"
                tooltip="Save changes"
                success
                m={2}
                fontSize={[3, 2]}
                show={show}
                plaintext
                primary
                onClick={this.handleOnEditClick}
                iconElement={<FaCheck />}
              />
            ),
            isUpdating && (
              <ActionButton
                key="SaveChanges"
                success
                m={2}
                fontSize={[3, 4]}
                show={show}
                plaintext
                primary
                onClick={this.handleOnEditClick}
                iconElement={<FaUpdating />}
              />
            ),
          ]) : (
            <Txt.p>
              {name}
            </Txt.p>
          )}
        </Flex>
      </CellContainer>
    );
  }
}

RoomCell.propTypes = {
  value: PropTypes.string.isRequired, // eslint-disable-line
  currentDevice: PropTypes.shape({ // eslint-disable-line
    name: PropTypes.string.isRequired,
  }),
  isUpdating: PropTypes.bool.isRequired,
  updateDeviceName: PropTypes.func.isRequired,
};

export default connect(mapStateToProps, mapDispatchToProps)(RoomCell);
