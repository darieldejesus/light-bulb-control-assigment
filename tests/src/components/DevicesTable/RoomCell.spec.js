import React from 'react';
import { Provider as RenditionProvider } from 'rendition';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider as ReduxProvider } from 'react-redux';

import api from '../../../../src/helpers/ApiClient';
import clientMiddleware from '../../../../src/stateManager/middleware/client';
import RoomCell, { RoomCell as RoomCellClass } from '../../../../src/components/DevicesTable/RoomCell';
import theme from '../../../../src/theme';

jest.mock('lodash/debounce', () => jest.fn(fn => fn));

const middlewares = [
  clientMiddleware(api),
  thunk,
];
const mockStore = configureMockStore(middlewares);

describe('RoomCell', () => {
  let props;
  let wrapper;
  const store = mockStore({
    isUpdatingName: false,
  });

  beforeEach(() => {
    props = {
      value: 'This is a name',
    };
    wrapper = shallow(
      <ReduxProvider store={store}>
        <RenditionProvider theme={theme}>
          <RoomCell {...props} />
        </RenditionProvider>
      </ReduxProvider>,
    );
  });

  test('Should render the component', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
