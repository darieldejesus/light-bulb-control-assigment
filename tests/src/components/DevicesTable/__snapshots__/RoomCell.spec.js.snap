// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoomCell Should render the component 1`] = `
<Provider
  theme={
    Object {
      "breakpoints": Array [
        32,
        48,
        64,
        80,
      ],
      "colors": Object {
        "danger": Object {
          "dark": "#661B1B",
          "light": "#FFECEC",
          "main": "#FF4444",
          "semilight": "#FFA1A1",
        },
        "gray": Object {
          "dark": "#9f9f9f",
          "light": "#f4f4f4",
          "main": "#c6c8c9",
        },
        "info": Object {
          "dark": "#0D3C58",
          "light": "#E9F4FB",
          "main": "#2297DE",
          "semilight": "#90CBEE",
        },
        "primary": Object {
          "dark": "#1E87C7",
          "light": "#5FB3E7",
          "main": "#2297DE",
        },
        "quartenary": Object {
          "dark": "#D3D6DB",
          "light": "#F0F3F7",
          "main": "#EBEFF4",
        },
        "secondary": Object {
          "dark": "#E5B01F",
          "light": "#FFD45E",
          "main": "#FFC523",
        },
        "statusConfiguring": Object {
          "main": "#ffb25e",
        },
        "statusIdle": Object {
          "main": "#89c683",
        },
        "statusOffline": Object {
          "main": "#fd7c7c",
        },
        "statusPostProvisioning": Object {
          "main": "#aa96d5",
        },
        "statusUpdating": Object {
          "main": "#7ccdfd",
        },
        "success": Object {
          "dark": "#1E4707",
          "light": "#EDF7E7",
          "main": "#4DB313",
          "semilight": "#A6D989",
        },
        "tertiary": Object {
          "dark": "#2F3033",
          "light": "#7E8085",
          "main": "#4D5057",
        },
        "text": Object {
          "dark": "#1c1d1e",
          "light": "#8f9297",
          "main": "#3c3e42",
        },
        "warning": Object {
          "dark": "#64410D",
          "light": "#FEF5E9",
          "main": "#FCA321",
          "semilight": "#FDD190",
        },
      },
      "fontSizes": Array [
        12,
        14,
        16,
        20,
        24,
        32,
        48,
        64,
        72,
        96,
      ],
      "radius": 3,
      "space": Array [
        0,
        4,
        8,
        16,
        36,
        48,
        128,
      ],
      "weights": Array [
        400,
        700,
      ],
    }
  }
>
  <Connect(RoomCell)
    value="This is a name"
  />
</Provider>
`;
